types/react-select/src/styles.d.ts

```diff
@@ -1,43 +1,53 @@
 import { CSSProperties } from 'react';
+import { ContainerProps, IndicatorContainerProps, ValueContainerProps } from './components/containers';
 import { ControlProps } from './components/Control';
+import { GroupProps, GroupHeadingProps } from './components/Group';
+import { IndicatorProps, LoadingIndicatorProps } from './components/indicators';
+import { MultiValueProps, MultiValueRemoveProps } from './components/MultiValue';
+import { InputProps } from './components/Input';
+import { MenuProps, MenuListComponentProps, MenuPortalProps, NoticeProps } from './components/Menu';
+import { OptionProps } from './components/Option';
 import { PlaceholderProps } from './components/Placeholder';
+import { SingleValueProps } from './components/SingleValue';
+import { OptionTypeBase } from './types';
 
+/** @deprecated - Used internally, but will not be exported in next major version */
 export interface Props { [key: string]: any; }
 
+/** @deprecated - Unused and will not be exported in next major version */
 export type StylesConfigFunction<Props = any> = (base: CSSProperties, props: Props) => CSSProperties;
 
 export interface Styles {
-  clearIndicator?: StylesConfigFunction;
-  container?: StylesConfigFunction;
-  control?: StylesConfigFunction<ControlProps<{}, boolean>>;
-  dropdownIndicator?: StylesConfigFunction;
-  group?: StylesConfigFunction;
-  groupHeading?: StylesConfigFunction;
-  indicatorsContainer?: StylesConfigFunction;
-  indicatorSeparator?: StylesConfigFunction;
-  input?: StylesConfigFunction;
-  loadingIndicator?: StylesConfigFunction;
-  // TODO loadingMessageCSS?: StylesConfigFunction;
-  loadingMessage?: StylesConfigFunction;
-  menu?: StylesConfigFunction;
-  menuList?: StylesConfigFunction;
-  menuPortal?: StylesConfigFunction;
-  multiValue?: StylesConfigFunction;
-  multiValueLabel?: StylesConfigFunction;
-  multiValueRemove?: StylesConfigFunction;
-  // TODO noOptionsMessageCSS?: StylesConfigFunction;
-  noOptionsMessage?: StylesConfigFunction;
-  option?: StylesConfigFunction;
-  placeholder?: StylesConfigFunction<PlaceholderProps<{}, boolean>>;
-  singleValue?: StylesConfigFunction;
-  valueContainer?: StylesConfigFunction;
+  clearIndicator?<OptionType extends OptionTypeBase, IsMulti extends boolean>(base: CSSProperties, props: IndicatorProps<OptionType, IsMulti>): CSSProperties;
```

Comment:
Sorry about the ugliness of this, but I found no way to use a generic in a generic 😢 

We could potentially leverage a generic in `Styles`, but this felt better despite the terseness - there already exist keys (`input`) that don't need the generic, and who is to say that wont continue to occur? Subjectively, in my opinion, a generic for `Style` would only feel right if it were used in every key.